### PR TITLE
Update version prompt if l2 binary below v1.5.1

### PR DESCRIPTION
**Pull request by lovestaco**
## What type of MR is this?

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Doc Update
- [ ] 🎨 Style
- [x] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] ⏩ Revert

## Description

- This MR adds the feature of Adding a version checker for v1.5.1 using _semver_.
- Update prompt for the l2 binary to the latest using `l2 -u`, prompt will come as a warning + description + button .
- Adding Terminal functionalities in Utils and using it in Execute and Gen Snippet.

### If there is no l2binary installed
image

### If l2binary is less than v1.5.1
image

image

## Important code file to start Code Review from
`Check l2Version.ts`

## Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [x] 🙋 no, because I need help
 
## Are there any post-deployment tasks we need to perform?

- Need to prettify the codebase before merging.

### CODE REVIEW

File: src/checkL2Version.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  } catch (e: any) {
+    console.log("Problem while checking for version -> ", e);
+  }
+}
+
+function getL2Version(callback: (version: string) => void) {
+  exec("l2 --version", (error, stdout, stderr) => {
+    const versionOutput = stdout.trim();
+    const versionMatch = versionOutput.match(/v(\d+\.\d+\.\d+)/);
+    if (versionMatch && versionMatch.length === 2) {
```
Comment (shrsv) [inline src/checkL2Version.ts]: Why `length === 2`?  We can use capture group to get a triple for (major, minor, patch) as well directly, I think.

File: src/checkL2Version.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      const l2Version = versionMatch[1];
+      callback(l2Version);
+    }
+  });
+}
+
+function compareL2Versions(currentVersion: string, latestVersion: string) {
```
Comment (shrsv) [inline src/checkL2Version.ts]: I recommend using a library suggested in this [SO Answer](https://stackoverflow.com/a/30256597). Our semver definition is simple now, but there are more changes possible. Best to simply use standardized library for it.

Should be as simple as:

```
semver.gte('3.4.8', '3.4.7') //true
```

Comment (shrsv) [reply src/checkL2Version.ts]: If we're using this library, then ignore the previous comment about regex*.

**Comment by shrsv** [comment]
Just curious - have we already integrated the variable suggestion logic (from both `l2.env` and `l2config.env`)? If not - we can do it in this MR itself.

File: src/checkL2Version.ts (hunk 1 above)
Comment (shrsv) [reply src/checkL2Version.ts]: I think we can try to use the official semver regex. See [here](https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string) for details. Check the actual regex in action at [regex101](https://regex101.com/r/vkijKf/1/)

File: src/extension.ts [hunk 3]
```diff
@@ -34,6 +35,9 @@ export function activate(context: vscode.ExtensionContext) {
   context.subscriptions.push(generateCodeSnippetDisposable);
   console.log(">>> generateCodeSnippetDisposable is now active!");
 
+  // Automatically check L2 version on extension activation
+  checkL2Version();
```
Comment (shrsv) [inline src/extension.ts]: So if version is too old, what happens to the variable  suggestion? Is it totally dead in older versions?

File: src/checkL2Version.ts (hunk 1 above)
Comment (lovestaco) [reply src/checkL2Version.ts]: Damn never knew so many kinds are there

File: src/checkL2Version.ts (hunk 2 above)
Comment (lovestaco) [reply src/checkL2Version.ts]: Yeah using this

**Comment by lovestaco** [comment]
> Just curious - have we already integrated the variable suggestion logic (from both `l2.env` and `l2config.env`)? If not - we can do it in this MR itself.

Yes its integrated to main but havent released it yet

[https://github.com/HexmosTech/Lama2Code/pull/10](https://github.com/HexmosTech/Lama2Code/pull/10)

File: src/extension.ts (hunk 3 above)
Comment (lovestaco) [reply src/extension.ts]: No in the older version what we were doing is searching the le.env file in front end and suggesting it. So it won't be a problem.

Comment (shrsv) [reply src/extension.ts]: What I am saying is something else.

Say, the extension frontend updates, but the l2 version is still old, what happens?

Remember - VSCode automatically updates extensions quite quickly (I think)

File: src/checkL2Version.ts [hunk 4]
```diff
@@ -1,62 +1,35 @@
...
+import * as semver from "semver";
 import * as vscode from "vscode";
 import { getShowLama2Term } from "./utils";
 
-const MIN_VERSION_TO_CHECK = "1.5.2";
 const LAMA2_TERM_NAME = "AutoLama2";
 const UPDATE_MSG = "Support for environment variables.";
 
-export function checkL2Version() {
+export async function getL2VersionAndUpdatePrompt(minVersionToCheck: string) {
```
Comment (shrsv) [inline src/checkL2Version.ts]: Which function call within this requires `async`? I don't see any `await`

File: src/checkL2Version.ts [hunk 5]
```diff
@@ -1,62 +1,35 @@
...
+    const normalizedVersion = semver.valid(l2Version);
+    if (normalizedVersion) {
+
+      if (semver.lt(l2Version, minVersionToCheck)) {
+        showUpdateWarning(minVersionToCheck);
       }
-    });
+      return normalizedVersion;
+    } else {
+      throw new Error("Invalid version format: " + l2Version);
```
Comment (shrsv) [inline src/checkL2Version.ts:20]: Shall we report to the user - that the version is unknown and probably the extension will not work in full? 

We can have a button "Download" to direct them to website.

File: src/checkL2Version.ts [hunk 6]
```diff
@@ -1,62 +1,35 @@
...
+    }
   } catch (e: any) {
     console.log("Problem while checking for version -> ", e);
```
Comment (shrsv) [inline src/checkL2Version.ts:23]: We should surface this as a warning - I think. 

Especially - what happens when *there is no binary* in the path? 

I think the user *ought to know* that something is wrong, and that action is needed.

Comment (lovestaco) [reply src/checkL2Version.ts:23]: Yeah added

File: src/checkL2Version.ts (hunk 5 above)
Comment (lovestaco) [reply src/checkL2Version.ts:20]: Added this

File: src/checkL2Version.ts (hunk 4 above)
Comment (lovestaco) [reply src/checkL2Version.ts]: Removed this, it wasn't needed

**Comment by shrsv** [comment]
LGTM

**Pull request merged**